spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -58,53 +61,138 @@ case class RewriteMergeInto(conf: SQLConf) extends Rule[LogicalPlan] with Rewrit
+
+        // use a full outer join because there are both matched and not matched actions
         val sourceTableProj = source.output ++ Seq(Alias(TRUE_LITERAL, ROW_FROM_SOURCE)())
-        val targetTableProj = target.output ++ Seq(Alias(TRUE_LITERAL, ROW_FROM_TARGET)())
+        val targetTableProj = targetTableScan.output ++ Seq(Alias(TRUE_LITERAL, ROW_FROM_TARGET)())
         val newTargetTableScan = Project(targetTableProj, targetTableScan)
         val newSourceTableScan = Project(sourceTableProj, source)
         val joinPlan = Join(newSourceTableScan, newTargetTableScan, FullOuter, Some(cond), JoinHint.NONE)
 
-        // Construct the plan to replace the data based on the output of `MergeInto`
         val mergeParams = MergeIntoParams(
           isSourceRowNotPresent = IsNull(findOutputAttr(joinPlan.output, ROW_FROM_SOURCE)),
           isTargetRowNotPresent = IsNull(findOutputAttr(joinPlan.output, ROW_FROM_TARGET)),
-          matchedConditions = matchedActions.map(getClauseCondition),
-          matchedOutputs = matchedActions.map(actionOutput(_, target.output)),
+          matchedConditions = matchedConditions,
+          matchedOutputs = matchedOutputs,
           notMatchedConditions = notMatchedActions.map(getClauseCondition),
-          notMatchedOutputs = notMatchedActions.map(actionOutput(_, target.output)),
-          targetOutput = target.output :+ FALSE_LITERAL,
-          deleteOutput = target.output :+ TRUE_LITERAL,
+          notMatchedOutputs = notMatchedActions.map(actionOutput),
+          targetOutput = target.output,
           joinedAttributes = joinPlan.output
         )
         val mergePlan = MergeInto(mergeParams, target, joinPlan)
         val batchWrite = mergeBuilder.asWriteBuilder.buildForBatch()
+
         ReplaceData(target, batchWrite, mergePlan)
     }
   }
 
-  private def actionOutput(clause: MergeAction, targetOutputCols: Seq[Expression]): Seq[Expression] = {
+  private def actionOutput(clause: MergeAction): Option[Seq[Expression]] = {
     clause match {
       case u: UpdateAction =>
-        u.assignments.map(_.value) :+ FALSE_LITERAL
+        Some(u.assignments.map(_.value))
       case _: DeleteAction =>
-        targetOutputCols :+ TRUE_LITERAL
+        None
       case i: InsertAction =>
-        i.assignments.map(_.value) :+ FALSE_LITERAL
+        Some(i.assignments.map(_.value))
     }
   }
 
   private def getClauseCondition(clause: MergeAction): Expression = {
     clause.condition.getOrElse(TRUE_LITERAL)
   }
+
+  private def buildDynamicFilterTargetScan(
+      target: DataSourceV2Relation,
+      source: LogicalPlan,
+      cond: Expression): (MergeBuilder, LogicalPlan) = {
+    // Construct the plan to prune target based on join condition between source and target.
+    val mergeBuilder = target.table.asMergeable.newMergeBuilder("merge", newWriteInfo(target.schema))
+    val matchingRowsPlanBuilder = (rel: DataSourceV2ScanRelation) =>
```

Comment:
I think it can fit on one line:

```
relation => Join(source, relation, Inner, Some(cond), JoinHint.NONE)
```

Reply:
I have this in my branch. If we need to fix anything major, I'll push it.